Spark 3.3: Split SparkScan and SparkBatch

aokolnychyi opened the pull request:
This PR splits `SparkScan` (logical representation) and `SparkBatch` (physical representation) into multiple classes and implements proper equality in `SparkBatch` so that `BatchScanExec` would work as expected in Spark.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatch.java:
```diff
@@ -115,18 +119,37 @@ private boolean parquetBatchReadsEnabled() {
   }
+
+  @Override
+  public boolean equals(Object o) {
```
@bryanck, could you, please, re-run the benchmark in which you spotted the performance regression? I think the current logic should make `BatchScanExec` behave correctly.

bryanck replied:
I ran the benchmark and this change doesn't have any negative impact on performance, so LGTM!

bryanck replied:
One note, I backported this to Spark 3.2 and ran it with that, as Spark 3.3 has a performance regression (https://issues.apache.org/jira/browse/SPARK-40703)

aokolnychyi replied:
Thank you, @bryanck! It is nice to see Chao already looking into that. Let's see what was the root cause.

aokolnychyi commented:
@rdblue @flyrain @kbendick @szehon-ho, any thoughts on separating `SparkBatch` and `SparkScan` as it was before we noticed a performance regression?

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScan.java:
```diff
@@ -101,15 +104,17 @@ protected List<Expression> filterExpressions() {
   @Override
   public Batch toBatch() {
-    return this;
+    return new SparkBatch(sparkContext, table, readConf, tasks(), expectedSchema, hashCode());
```
In line with `SparkMicroBatchStream` below.

shardulm94 commented:
@aokolnychyi Can you give more context on what doesn't work as expected when `SparkBatchScan` is a single class? I may be missing some context discussed elsewhere.

aokolnychyi commented:
@shardulm94, I refactored `SparkScan` (logical representation) and `SparkBatch` (physical representation) into multiple classes some time back but that caused a performance regression as we did not implement proper equality for `SparkBatch`. Then we reverted that change in PR #5143 to avoid the performance issue. This PR changes it back and overrides equality in `SparkBatch` so that everything works as expected.

shardulm94 commented:
Thanks @aokolnychyi!

flyrain reviewed (approved):
+1

aokolnychyi commented:
Thanks for reviewing everyone!
